File: scripts/deploy.py
"""
Deploy all 5 Unsubscribely escrow vault contracts to Algorand Testnet.

Usage:
    python scripts/deploy.py

Required environment variables (.env file):
    TESTNET_MNEMONIC   — 25-word mnemonic of the deployer account (must be funded)
    ALGOD_URL          — (optional) defaults to https://testnet-api.algonode.cloud
    ALGOD_TOKEN        — (optional) defaults to empty string for AlgoNode public API
"""

import os
import json
import sys
from pathlib import Path

import algosdk
from algosdk.v2client import algod
from algosdk import transaction, account, encoding, logic
from dotenv import load_dotenv

load_dotenv()

ALGOD_URL = os.getenv("ALGOD_URL", "https://testnet-api.algonode.cloud")
ALGOD_TOKEN = os.getenv("ALGOD_TOKEN", "")
TESTNET_MNEMONIC = os.getenv("TESTNET_MNEMONIC", "")

ARTIFACTS_DIR = Path(__file__).parent.parent / "smart_contracts" / "artifacts"
DEPLOYMENT_RESULTS_FILE = Path(__file__).parent.parent / "smart_contracts" / "artifacts" / "deployed.json"

CONTRACT_CONFIGS = [
    {
        "name": "EscrowVault",
        "global_schema": transaction.StateSchema(num_uints=1, num_byte_slices=2),
        "local_schema": transaction.StateSchema(0, 0),
        "description": "Standard Escrow Vault",
    },
    {
        "name": "TimeLockEscrow",
        "global_schema": transaction.StateSchema(num_uints=2, num_byte_slices=2),
        "local_schema": transaction.StateSchema(0, 0),
        "description": "Time-Locked Escrow Vault",
    },
    {
        "name": "MultiSigEscrow",
        "global_schema": transaction.StateSchema(num_uints=3, num_byte_slices=3),
        "local_schema": transaction.StateSchema(0, 0),
        "description": "Multi-Signature Escrow Vault",
    },
    {
        "name": "DisputeEscrow",
        "global_schema": transaction.StateSchema(num_uints=1, num_byte_slices=3),
        "local_schema": transaction.StateSchema(0, 0),
        "description": "Dispute-Resolution Escrow Vault",
    },
    {
        "name": "ASAEscrow",
        "global_schema": transaction.StateSchema(num_uints=2, num_byte_slices=2),
        "local_schema": transaction.StateSchema(0, 0),
        "description": "ASA Token Escrow Vault",
    },
]


def get_algod_client() -> algod.AlgodClient:
    return algod.AlgodClient(ALGOD_TOKEN, ALGOD_URL)


def load_teal(contract_name: str, kind: str) -> bytes:
    """Load compiled TEAL bytes from artifacts directory.
    Supports both naming conventions: '{kind}.teal' and '{name}.{kind}.teal'
    """
    candidates = [
        ARTIFACTS_DIR / contract_name / f"{contract_name}.{kind}.teal",
        ARTIFACTS_DIR / contract_name / f"{kind}.teal",
    ]
    for path in candidates:
        if path.exists():
            return path.read_bytes()
    raise FileNotFoundError(
        f"TEAL file not found. Tried:\n" +
        "\n".join(f"  {p}" for p in candidates) +
        "\nRun 'algokit compile python smart_contracts/' to generate artifacts."
    )


def compile_teal(algod_client: algod.AlgodClient, teal_source: bytes) -> bytes:
    """Compile TEAL source to binary via algod."""
    result = algod_client.compile(teal_source.decode())
    import base64
    return base64.b64decode(result["result"])


def deploy_contract(
    algod_client: algod.AlgodClient,
    creator_address: str,
    creator_private_key: str,
    approval_program: bytes,
    clear_program: bytes,
    global_schema: transaction.StateSchema,
    local_schema: transaction.StateSchema,
) -> int:
    """Deploy an application and return its app ID."""
    params = algod_client.suggested_params()

    txn = transaction.ApplicationCreateTxn(
        sender=creator_address,
        sp=params,
        on_complete=transaction.OnComplete.NoOpOC,
        approval_program=approval_program,
        clear_program=clear_program,
        global_schema=global_schema,
        local_schema=local_schema,
    )

    signed = txn.sign(creator_private_key)
    tx_id = algod_client.send_transaction(signed)
    print(f"    Transaction ID: {tx_id}")

    result = transaction.wait_for_confirmation(algod_client, tx_id, 4)
    return result["application-index"]


def main():
    if not TESTNET_MNEMONIC:
        print("ERROR: TESTNET_MNEMONIC environment variable is not set.")
        print("Create a .env file with your 25-word mnemonic:")
        print('  TESTNET_MNEMONIC="word1 word2 ... word25"')
        sys.exit(1)

    private_key = algosdk.mnemonic.to_private_key(TESTNET_MNEMONIC)
    deployer_address = account.address_from_private_key(private_key)

    client = get_algod_client()

    account_info = client.account_info(deployer_address)
    balance_algo = account_info.get("amount", 0) / 1_000_000
    print(f"Deployer: {deployer_address}")
    print(f"Balance:  {balance_algo:.6f} ALGO")
    print(f"Network:  {ALGOD_URL}")
    print("=" * 60)

    if balance_algo < 1.0:
        print("WARNING: Low balance. Fund via https://bank.testnet.algorand.network/")

    deployed = {}

    for config in CONTRACT_CONFIGS:
        name = config["name"]
        print(f"\nDeploying {name} ({config['description']})...")

        try:
            approval_teal = load_teal(name, "approval")
            clear_teal = load_teal(name, "clear")

            approval_binary = compile_teal(client, approval_teal)
            clear_binary = compile_teal(client, clear_teal)

            app_id = deploy_contract(
                client,
                deployer_address,
                private_key,
                approval_binary,
                clear_binary,
                config["global_schema"],
                config["local_schema"],
            )

            app_address = logic.get_application_address(app_id)

            deployed[name] = {
                "app_id": app_id,
                "app_address": app_address,
                "network": ALGOD_URL,
                "deployer": deployer_address,
            }

            print(f"    App ID:      {app_id}")
            print(f"    App Address: {app_address}")

        except FileNotFoundError as e:
            print(f"    SKIPPED — {e}")
            deployed[name] = {"error": str(e)}
        except Exception as e:
            print(f"    FAILED — {e}")
            deployed[name] = {"error": str(e)}

    print("\n" + "=" * 60)
    print("Deployment Summary:")
    for name, info in deployed.items():
        if "app_id" in info:
            print(f"  {name}: App ID {info['app_id']} @ {info['app_address']}")
        else:
            print(f"  {name}: {info.get('error', 'unknown error')}")

    DEPLOYMENT_RESULTS_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(DEPLOYMENT_RESULTS_FILE, "w") as f:
        json.dump(deployed, f, indent=2)

    print(f"\nDeployment results saved to: {DEPLOYMENT_RESULTS_FILE}")


if __name__ == "__main__":
    main()

File: scripts/deploy.js
#!/usr/bin/env node
/**
 * Deploy all 5 Unsubscribely escrow vault contracts to Algorand Testnet.
 *
 * Usage:
 *   node scripts/deploy.js
 *
 * Required environment variables (.env file):
 *   TESTNET_MNEMONIC   — 25-word mnemonic of a funded Testnet account
 *   ALGOD_URL          — (optional) defaults to https://testnet-api.algonode.cloud
 *   ALGOD_TOKEN        — (optional) defaults to empty string
 *
 * Fund your deployer account at: https://bank.testnet.algorand.network/
 */

import algosdk from "algosdk"
import { readFileSync, writeFileSync, existsSync } from "fs"
import { join, dirname } from "path"
import { fileURLToPath } from "url"

const __dirname = dirname(fileURLToPath(import.meta.url))
const ROOT = join(__dirname, "..")
const ARTIFACTS_DIR = join(ROOT, "smart_contracts", "artifacts")
const OUTPUT_FILE = join(ARTIFACTS_DIR, "deployed.json")

const ALGOD_URL = process.env.ALGOD_URL || "https://testnet-api.algonode.cloud"
const ALGOD_TOKEN = process.env.ALGOD_TOKEN || ""
const TESTNET_MNEMONIC = process.env.TESTNET_MNEMONIC || ""

const CONTRACT_CONFIGS = [
  {
    name: "EscrowVault",
    description: "Standard Escrow Vault",
    numGlobalInts: 1, numGlobalBytes: 2,
    numLocalInts: 0, numLocalBytes: 0,
  },
  {
    name: "AgentEscrowVault",
    description: "Agent-Managed Escrow Vault (A2A Autonomous Payments)",
    numGlobalInts: 1, numGlobalBytes: 3,
    numLocalInts: 0, numLocalBytes: 0,
  },
  {
    name: "TimeLockEscrow",
    description: "Time-Locked Escrow Vault",
    numGlobalInts: 2, numGlobalBytes: 2,
    numLocalInts: 0, numLocalBytes: 0,
  },
  {
    name: "MultiSigEscrow",
    description: "Multi-Signature Escrow Vault",
    numGlobalInts: 3, numGlobalBytes: 3,
    numLocalInts: 0, numLocalBytes: 0,
  },
  {
    name: "DisputeEscrow",
    description: "Dispute-Resolution Escrow Vault",
    numGlobalInts: 1, numGlobalBytes: 3,
    numLocalInts: 0, numLocalBytes: 0,
  },
  {
    name: "ASAEscrow",
    description: "ASA Token Escrow Vault",
    numGlobalInts: 2, numGlobalBytes: 2,
    numLocalInts: 0, numLocalBytes: 0,
  },
  // ── Singleton contracts (one app for the whole platform) ──────────────────
  {
    name: "ServiceRegistry",
    description: "A2A on-chain service registry (Box-Storage)",
    numGlobalInts: 2, numGlobalBytes: 1,
    numLocalInts: 0, numLocalBytes: 0,
    singleton: true,
  },
  {
    name: "AgentEscrowVaultV2",
    description: "Agent Escrow v2 with on-chain billing history (Box-Storage)",
    numGlobalInts: 2, numGlobalBytes: 3,
    numLocalInts: 0, numLocalBytes: 0,
    template: true, // deploys one canonical instance for spec discovery only
  },
]

function loadTeal(name, kind) {
  const candidates = [
    join(ARTIFACTS_DIR, name, `${name}.${kind}.teal`),
    join(ARTIFACTS_DIR, name, `${kind}.teal`),
  ]
  for (const path of candidates) {
    if (existsSync(path)) {
      return readFileSync(path)
    }
  }
  throw new Error(`TEAL file not found for ${name}/${kind}. Tried:\n${candidates.join("\n")}`)
}

async function compileTeal(algodClient, tealSource) {
  const result = await algodClient.compile(tealSource).do()
  return Buffer.from(result.result, "base64")
}

function loadArc56(name) {
  const path = join(ARTIFACTS_DIR, name, `${name}.arc56.json`)
  if (!existsSync(path)) return null
  return JSON.parse(readFileSync(path, "utf-8"))
}

function buildCreateAppArgs(arc56, deployerAddress) {
  // Find the ABI method that has create=NoOp action
  const createMethod = arc56?.methods?.find(m => (m.actions?.create || []).includes("NoOp"))
  if (!createMethod) return [] // bare create

  const sigArgs = createMethod.args.map(a => a.type).join(",")
  const signature = `${createMethod.name}(${sigArgs})${createMethod.returns?.type || "void"}`
  const selector = new algosdk.ABIMethod({
    name: createMethod.name,
    args: createMethod.args,
    returns: createMethod.returns || { type: "void" },
  }).getSelector()

  const encoded = createMethod.args.map(a => {
    const t = algosdk.ABIType.from(a.type)
    // Sensible defaults: addresses → deployer; uint64 → 0
    if (a.type === "address") return t.encode(deployerAddress)
    if (a.type.startsWith("uint")) return t.encode(BigInt(0))
    if (a.type === "string") return t.encode("")
    if (a.type === "bool") return t.encode(false)
    return t.encode(BigInt(0))
  })

  console.log(`    ABI create: ${signature}`)
  return [selector, ...encoded]
}

async function deployContract(algodClient, senderAddress, senderSk, approvalBinary, clearBinary, config, arc56) {
  const suggestedParams = await algodClient.getTransactionParams().do()
  const appArgs = buildCreateAppArgs(arc56, senderAddress)

  const txn = algosdk.makeApplicationCreateTxnFromObject({
    sender: senderAddress,
    suggestedParams,
    onComplete: algosdk.OnApplicationComplete.NoOpOC,
    approvalProgram: approvalBinary,
    clearProgram: clearBinary,
    numGlobalByteSlices: config.numGlobalBytes,
    numGlobalInts: config.numGlobalInts,
    numLocalByteSlices: config.numLocalBytes,
    numLocalInts: config.numLocalInts,
    appArgs: appArgs.length ? appArgs : undefined,
  })

  const signedTxn = txn.signTxn(senderSk)
  const { txid } = await algodClient.sendRawTransaction(signedTxn).do()
  console.log(`    TxID: ${txid}`)

  const result = await algosdk.waitForConfirmation(algodClient, txid, 4)
  const appIdx = result["application-index"] ?? result.applicationIndex
  return Number(appIdx)
}

async function main() {
  if (!TESTNET_MNEMONIC) {
    console.error("ERROR: TESTNET_MNEMONIC environment variable is not set.")
    console.error("Set it in your .env file with your 25-word mnemonic.")
    process.exit(1)
  }

  const privateKey = algosdk.mnemonicToSecretKey(TESTNET_MNEMONIC)
  const senderAddress = privateKey.addr.toString()
  const senderSk = privateKey.sk

  const algodClient = new algosdk.Algodv2(ALGOD_TOKEN, ALGOD_URL, "")

  const accountInfo = await algodClient.accountInformation(senderAddress).do()
  const balanceAlgo = Number(accountInfo.amount) / 1_000_000
  console.log(`Deployer : ${senderAddress}`)
  console.log(`Balance  : ${balanceAlgo.toFixed(6)} ALGO`)
  console.log(`Network  : ${ALGOD_URL}`)
  console.log("=".repeat(60))

  if (balanceAlgo < 1.0) {
    console.warn(`\nWARNING: Low balance (${balanceAlgo.toFixed(4)} ALGO).`)
    console.warn(`Fund this address: https://bank.testnet.algorand.network/`)
    console.warn(`Address: ${senderAddress}\n`)
    if (balanceAlgo === 0) {
      console.error("ERROR: Zero balance — cannot deploy. Fund the account first.")
      process.exit(1)
    }
  }

  const deployed = {}

  // Only deploy SINGLETON contracts here. Per-user vaults are created by each
  // user from the UI signed by their own wallet.
  const ONLY_SINGLETONS = process.env.DEPLOY_ALL !== "1"
  const targets = ONLY_SINGLETONS
    ? CONTRACT_CONFIGS.filter(c => c.singleton || c.template)
    : CONTRACT_CONFIGS

  console.log(`Deploying ${targets.length} contract(s)${ONLY_SINGLETONS ? " (singletons only — set DEPLOY_ALL=1 for all)" : ""}\n`)

  for (const config of targets) {
    console.log(`\nDeploying ${config.name} (${config.description})...`)
    try {
      const approvalTeal = loadTeal(config.name, "approval")
      const clearTeal = loadTeal(config.name, "clear")
      const arc56 = loadArc56(config.name)

      const approvalBinary = await compileTeal(algodClient, approvalTeal)
      const clearBinary = await compileTeal(algodClient, clearTeal)

      const appId = await deployContract(algodClient, senderAddress, senderSk, approvalBinary, clearBinary, config, arc56)
      const appAddrObj = algosdk.getApplicationAddress(appId)
      const appAddress = typeof appAddrObj === "string"
        ? appAddrObj
        : algosdk.encodeAddress(appAddrObj.publicKey)

      deployed[config.name] = {
        app_id: appId,
        app_address: appAddress,
        lora_url: `https://lora.algokit.io/testnet/application/${appId}`,
        network: ALGOD_URL,
        deployer: senderAddress,
      }

      console.log(`    App ID:      ${appId}`)
      console.log(`    App Address: ${appAddress}`)
      console.log(`    Lora URL:    https://lora.algokit.io/testnet/application/${appId}`)
    } catch (err) {
      console.error(`    FAILED — ${err.message}`)
      deployed[config.name] = { error: err.message }
    }
  }

  console.log("\n" + "=".repeat(60))
  console.log("Deployment Summary:")
  for (const [name, info] of Object.entries(deployed)) {
    if (info.app_id) {
      console.log(`  ${name}: App ID ${info.app_id}`)
      console.log(`    Lora: ${info.lora_url}`)
    } else {
      console.log(`  ${name}: ERROR — ${info.error}`)
    }
  }

  writeFileSync(OUTPUT_FILE, JSON.stringify(deployed, null, 2))
  console.log(`\nDeployment results saved to: ${OUTPUT_FILE}`)
}

main().catch((err) => {
  console.error("Fatal:", err.message)
  process.exit(1)
})
